pyswarms/single/general_optimizer.py

```diff
@@ -174,21 +182,38 @@ def __init__(
                     "or 2 (for L2/Euclidean)."
                 )
 
-        # Case for Random and Ring topologies
-        if isinstance(topology, (Random, Ring)):
-            # Assign k-neighbors as attribute
-            self.k = options["k"]
-            if not isinstance(self.k, int):
-                raise ValueError(
-                    "No. of neighbors must be an integer between"
-                    "0 and no. of particles."
-                )
-            if not 0 <= self.k <= self.n_particles - 1:
-                raise ValueError(
-                    "No. of neighbors must be between 0 and no. " "of particles."
-                )
-            if "k" not in self.options:
-                raise KeyError("Missing k in options")
+        # Case for Random, VonNeumann and Ring topologies
+        if isinstance(topology, (Random, Ring, VonNeumann)):
+            if not isinstance(topology, VonNeumann):
+                self.k = options["k"]
+                if not isinstance(self.k, int):
+                    raise ValueError(
+                        "No. of neighbors must be an integer between"
+                        "0 and no. of particles."
+                    )
+                if not 0 <= self.k <= self.n_particles - 1:
+                    raise ValueError(
+                        "No. of neighbors must be between 0 and no. "
+                        "of particles."
+                    )
+                if "k" not in self.options:
+                    raise KeyError("Missing k in options")
+            else:
+                # Assign range r as attribute
+                self.r = options["r"]
+                if not isinstance(self.r, int):
+                    raise ValueError("The range must be a positive integer")
+                if (
+                    self.r <= 0
+                    or not 0
+                           <= VonNeumann.delannoy(self.swarm.dimensions, self.r)
+                           <= self.n_particles - 1
+                ):
+                    raise ValueError(
+                        "The range must be set such that the computed"
+                        "Delannoy number (number of neighbours) is"
+                        "between 0 and the no. of particles."
+                    )
```

Comment:
I'll accept this for now, but for v.0.4.0 we should make our if-statements shallower. In addition, it would be nice if we can create (for the next version) an `errors.py` module for our own custom exceptions. That would be nice. I'll put that in the pipeline.